python/tvm/contrib/graph_runtime.py

```diff
@@ -39,9 +38,113 @@ def create(graph_json_str, libmod, ctx):
         hmod = rpc_base._ModuleHandle(libmod)
         fcreate = ctx._rpc_sess.get_function("tvm.graph_runtime.remote_create")
         device_type = device_type % rpc_base.RPC_SESS_MASK
-        return GraphModule(fcreate(graph_json_str, hmod, device_type, device_id), ctx)
+        return GraphModule(fcreate(graph_json_str, hmod, device_type,
+                                   device_id), ctx)
+
     fcreate = get_global_func("tvm.graph_runtime.create")
-    return GraphModule(fcreate(graph_json_str, libmod, device_type, device_id), ctx)
+    return GraphModule(fcreate(graph_json_str, libmod, device_type,
+                               device_id), ctx)
+
+
+def _create_heterogeneous(graph_json_str, libmod_ctx, local_ctx):
+    """Create a heterogeneous runtime executor module given a graph and module.
+
+    Parameters
+    ----------
+    graph_json_str : str or graph class
+        The graph to be deployed in json format output by nnvm graph.
+        The graph can only contain tvm_op and device_copy_op that
+        point to the name of PackedFunc in the one of the compiled module libs.
+
+    libmod_ctx : tvm.Module to TVMContext dict
+        The module and context pair of the corresponding function.
+
+    local_ctx : TVMContext
+        The local context to deploy the module.
+
+    Returns
+    -------
+    graph_module : GraphModule
+        Runtime graph module that can be used to execute the graph.
+    """
+    if local_ctx.device_type >= rpc_base.RPC_SESS_MASK:
+        raise RuntimeError(
+            "rpc is not supported for heterogeneous execution yet.")
+
+    # Fallback to use the homogeneous execution if there is only one context.
+    if len(libmod_ctx) == 1:
+        return _create_homogeneous(graph_json_str, list(libmod_ctx.keys())[0],
+                                   list(libmod_ctx.values())[0])
+
+    libs, device_types, device_ids = [], [], []
+    # CPU is always assumed to be the host processor when deploying on a
+    # heterogeneous platform. For convenience, CPU related arguments are set as
+    # the firs elements when passing the module and ctx pairs to the backend.
+    for lib, ctx in libmod_ctx.items():
+        if ctx.device_type >= rpc_base.RPC_SESS_MASK:
+            raise RuntimeError(
+                "rpc is not supported for heterogeneous execution yet.")
+        if ctx == tvm.cpu(ctx.device_id):
```

Comment:
In this code segment, are we assuming that, considering the *subgraph* of the *whole graph*, each target device will only be used by one *subgraph* at most?

Reply:
Not really. The "subgraph" is actually just a fused node here. For example, we annotate the graph and provide the context information for each node. During fusion, nodes with different context will not be grouped together. Instead, different ops marked with the same device/context will be compiled into the same binary.